src/norm.jl

```diff
@@ -0,0 +1,92 @@
+# Code adapted from Flux.jl:
+# https://github.com/FluxML/Flux.jl/blob/68ba6e4e2fa4b86e2fef8dc6d0a5d795428a6fac/src/layers/normalise.jl#L117-L206
+# License: https://github.com/FluxML/Flux.jl/blob/master/LICENSE.md
+
+mutable struct BatchNormFlow{T1,T2} <: Bijector
+    β::T1
+    logγ::T1
+    μ  # moving mean
+    σ² # moving st
+    ϵ::T2
+    momentum::T2
+    active::Bool # true when training
+end
+
+BatchNormFlow(dims::Int, container=Array; ϵ=1f-5, momentum=0.1f0) = BatchNormFlow(
+    container(zeros(Float32, dims)),
+    container(zeros(Float32, dims)),
+    zeros(Float32, dims),
+    ones(Float32, dims),
+    ϵ,
+    momentum,
+    true
+)
+
+function affinesize(x)
+    dims = length(size(x))
+    channels = size(x, dims-1)
+    affinesize = ones(Int, dims)
+    affinesize[end-1] = channels
+    return affinesize
+end
+
+logabsdetjacob(
+    t::T,
+    x;
+    σ²=reshape(t.σ², affinesize(x)...)
+) where {T<:BatchNormFlow} =  (sum(t.logγ - log.(σ² .+ t.ϵ) / 2)) .* typeof(Flux.data(x))(ones(Float32, size(x, 2))')
```

Comment:
Does all of this work nicely with `Tracker`? I.e. we're not getting cases where `TrackedArray` is converted into `Array{<:TrackedReal}`.

Reply:
@torfjelde Could you give an example when this could happen?

Reply:
Should be fine with `Trakcer` considering it's adapted from Flux. Would be nice to provide some examples which could be potentiall problematic @torfjelde

Reply:
```julia
using Bijectors, Tracker

julia> x = param(ones(1, 2))
Tracked 1×2 Array{Float64,2}:
 1.0  1.0

julia> b = Bijectors.InvertibleBatchNorm(2, Tracker.param)
InvertibleBatchNorm{TrackedArray{…,Array{Float64,2}},Array{Float64,2},Float32}([0.0; 0.0] (tracked), [0.0; 0.0] (tracked), [0.0; 0.0], [1.0; 1.0], 1.0f-5, 0.1f0, true)

julia> logabsdetjac(b, x)
ERROR: MethodError: no method matching Float64(::Tracker.TrackedReal{Float64})
Closest candidates are:
  Float64(::Real, ::RoundingMode) where T<:AbstractFloat at rounding.jl:185
  Float64(::T<:Number) where T<:Number at boot.jl:725
  Float64(::Int8) at float.jl:60
  ...
Stacktrace:
 [1] convert(::Type{Float64}, ::Tracker.TrackedReal{Float64}) at ./number.jl:7
 [2] setindex! at ./array.jl:771 [inlined]
 [3] setindex! at ./multidimensional.jl:414 [inlined]
 [4] copyto!(::Array{Float64,2}, ::TrackedArray{…,Array{Float64,2}}) at ./multidimensional.jl:830
 [5] convert(::Type{Array{Float64,2}}, ::TrackedArray{…,Array{Float64,2}}) at ./array.jl:497
 [6] setproperty!(::InvertibleBatchNorm{TrackedArray{…,Array{Float64,2}},Array{Float64,2},Float32}, ::Symbol, ::TrackedArray{…,Array{Float64,2}}) at ./sysimg.jl:19
 [7] forward(::InvertibleBatchNorm{TrackedArray{…,Array{Float64,2}},Array{Float64,2},Float32}, ::TrackedArray{…,Array{Float64,2}}) at /scratch/tef30/.julia/dev/Bijectors/src/batch_norm.jl:87
 [8] logabsdetjac(::InvertibleBatchNorm{TrackedArray{…,Array{Float64,2}},Array{Float64,2},Float32}, ::TrackedArray{…,Array{Float64,2}}) at /scratch/tef30/.julia/dev/Bijectors/src/batch_norm.jl:53
 [9] top-level scope at none:0
```

Reply:
Changing the `active` field makes it run without issues (though the batch-shape should still be transposed):
```julia
julia> b.active = false
false

julia> logabsdetjac(b, x)
Tracked 1×1 Adjoint{Float64,Array{Float64,2}}:
 -9.999949747714609e-6
```

But I suspect it's supposed to work when `active == true` also, right?

Reply:
Oh, and the transform and everything related should also work with a single input rather than always requiring a batch. A quick fix is to dispatch differently on `AbstractVector` and `AbstractMatrix`

Reply:
Good point. I never tried this transform with a single vector but we should make it work.

Reply:
> Transform and logabsdetjac with single variable seems to work provided it is in the format (n_dims, batch_size) and it is not tracked. I think we should enforce the input to be (n_dims, batch_size) or another possible solution is adding x = reshape(x, :, 1) everywhere were x is taken as input.

I really don't think we want to enforce `(n_dims, batch_size)`. What I did for the other normalizing flows (they had the same problem), is that I just defined a `(b:Flow)(x::AbstractVector) = b(reshape(x, 1, :))`. You can just do this for the relevant functions, e.g. `logabsdetjac`, and we're good:)

> Do you think this will work for the tracked input problem?

I don't quite see what the purpose of that change is to be honest. From a quick look it doesn't actually seem to have a different effect than the old code. Have you tried this? I don't know exactly what the problem above is, as I haven't looked properly into it yet. I'd suggest looking at the lines referenced in the error and work from there.

Reply:
> ERROR: AssertionError: `InvertibleBatchNorm` expected 2 channels, got 1
This initial error which you had come across is because of the input not taken as column-major. Now, this shouldn't be a problem. 

After fixing this another problem occurs.
[This](https://github.com/TuringLang/Bijectors.jl/pull/43/commits/03a16592ac2de7d948200ad71794ecc2f5debef7) commit is to fix the following error.
```julia
julia> b = InvertibleBatchNorm(2)
InvertibleBatchNorm{Array{Float64,2},Array{Float64,2},Float32}([0.0; 0.0], [0.0; 0.0], [0.0; 0.0], [1.0; 1.0], 1.0f-5, 0.1f0, true)

julia> x = param(ones(2, 1)); b(x)
ERROR: MethodError: no method matching Float64(::Tracker.TrackedReal{Float64})
Closest candidates are:
  Float64(::Real, ::RoundingMode) where T<:AbstractFloat at rounding.jl:194
  Float64(::T<:Number) where T<:Number at boot.jl:741
  Float64(::Int8) at float.jl:60
  ...
Stacktrace:
 [1] convert(::Type{Float64}, ::Tracker.TrackedReal{Float64}) at ./number.jl:7
 [2] setindex! at ./array.jl:769 [inlined]
 [3] setindex! at ./multidimensional.jl:460 [inlined]
 [4] copyto!(::Array{Float64,2}, ::TrackedArray{…,Array{Float64,2}}) at ./multidimensional.jl:877
 [5] convert(::Type{Array{Float64,2}}, ::TrackedArray{…,Array{Float64,2}}) at ./array.jl:482
 [6] setproperty!(::InvertibleBatchNorm{Array{Float64,2},Array{Float64,2},Float32}, ::Symbol, ::TrackedArray{…,Array{Float64,2}}) at ./sysimg.jl:19
 [7] forward(::InvertibleBatchNorm{Array{Float64,2},Array{Float64,2},Float32}, ::TrackedArray{…,Array{Float64,2}}) at /media/sharan/Work/Coding/MachineLearning/Projects/julia/Bijectors.jl/src/batch_norm.jl:93
 [8] (::InvertibleBatchNorm{Array{Float64,2},Array{Float64,2},Float32})(::TrackedArray{…,Array{Float64,2}}) at /media/sharan/Work/Coding/MachineLearning/Projects/julia/Bijectors.jl/src/batch_norm.jl:106
 [9] top-level scope at none:0
```

Reply:
> Good point. I never tried this transform with a single vector but we should make it work.

OK I was wrong here. We never input a single vector to batch norm. Batch norm only works when input is a batch, i.e. more than a vector.

Reply:
> OK I was wrong here. We never input a single vector to batch norm. Batch norm only works when input is a batch, i.e. more than a vector.

Ah, true! Should maybe raise an explicit error for this? So that people like me don't make that mistake again:)